File: PITA-Taller1/python/services/payroll_engine.py
"""Pure payroll calculation domain for Python and C++ parity.

The engine receives immutable-like snapshots and returns a structured result. It
never reads persisted payroll values from Professor or Administrative and has no
GUI, persistence, or printing dependencies.
"""

from dataclasses import asdict, dataclass, field
from datetime import date
from decimal import Decimal, ROUND_HALF_UP
from typing import Iterable, Mapping


class PayrollDomainError(ValueError):
    """Raised when payroll input or configuration violates a domain rule."""


Money = Decimal
ZERO = Decimal("0")
ONE = Decimal("1")
DAYS_IN_YEAR = Decimal("360")


def money(value) -> Money:
    """Converts a numeric value to Decimal without inheriting binary float noise."""
    if isinstance(value, Decimal):
        return value
    return Decimal(str(value))


def round_peso(value) -> Money:
    return money(value).quantize(Decimal("1"), rounding=ROUND_HALF_UP)


def _non_negative(value, field_name):
    result = money(value)
    if result < ZERO:
        raise PayrollDomainError(f"{field_name} cannot be negative")
    return result


@dataclass(frozen=True)
class PayrollPeriod:
    period: str
    start_date: date
    end_date: date
    days_worked: int

    def validate(self):
        if not self.period.strip():
            raise PayrollDomainError("period is required")
        if self.end_date < self.start_date:
            raise PayrollDomainError("period end_date cannot precede start_date")
        if not 0 <= self.days_worked <= 360:
            raise PayrollDomainError("days_worked must be between 0 and 360")


@dataclass(frozen=True)
class PayrollEmployee:
    employee_id: str
    employee_type: str
    employment_type: str
    base_monthly_salary: Money = ZERO
    point_value: Money = ZERO
    category_score: Money = ZERO
    title_score: Money = ZERO
    experience_score: Money = ZERO
    productivity_score: Money = ZERO
    academic_management_score: Money = ZERO
    hourly_rate: Money = ZERO
    hours_worked: Money = ZERO
    continuous_service_days: int = 0
    active: bool = True
    hire_date: str = ""
    termination_date: str = ""

    def validate(self):
        if not str(self.employee_id).strip():
            raise PayrollDomainError("employee_id is required")
        if self.employee_type not in {"Professor", "Administrative"}:
            raise PayrollDomainError("unsupported employee_type")
        if self.days_fields_negative():
            raise PayrollDomainError("employee numeric values cannot be negative")
        if self.continuous_service_days < 0:
            raise PayrollDomainError("continuous_service_days cannot be negative")

    def days_fields_negative(self):
        return any(
            money(getattr(self, name)) < ZERO
            for name in (
                "base_monthly_salary", "point_value", "category_score", "title_score",
                "experience_score", "productivity_score", "academic_management_score",
                "hourly_rate", "hours_worked",
            )
        )

    @property
    def total_points(self):
        return sum(
            (money(getattr(self, name)) for name in (
                "category_score", "title_score", "experience_score",
                "productivity_score", "academic_management_score",
            )),
            ZERO,
        )


@dataclass(frozen=True)
class PayrollNovelty:
    code: str
    amount: Money = ZERO
    quantity: Money = ONE
    is_salary: bool = True
    affects_ibc: bool = True
    description: str = ""

    def validate(self):
        if not self.code.strip():
            raise PayrollDomainError("novelty code is required")
        _non_negative(self.amount, "novelty.amount")
        _non_negative(self.quantity, "novelty.quantity")


@dataclass(frozen=True)
class PayrollRules:
    employee_health_rate: Money = Decimal("0.04")
    employee_pension_rate: Money = Decimal("0.04")
    service_bonus_rate: Money = Decimal("0.0833")
    severance_rate: Money = Decimal("0.0833")
    severance_interest_rate: Money = Decimal("0.12")
    christmas_bonus_rate: Money = Decimal("0.0833")
    vacation_rate: Money = Decimal("0.0417")
    vacation_divisor: Money = Decimal("720")
    vacation_bonus_rate: Money = Decimal("0.0556")
    employer_pension_rate: Money = Decimal("0.12")
    employer_health_rate: Money = Decimal("0.085")
    employer_health_exempt_rate: Money = ZERO
    arl_rates: Mapping[str, Money] = field(default_factory=lambda: {"I": Decimal("0.00522")})
    compensation_fund_rate: Money = Decimal("0.04")
    sena_rate: Money = Decimal("0.02")
    sena_exempt_rate: Money = ZERO
    icbf_rate: Money = Decimal("0.03")
    icbf_exempt_rate: Money = ZERO
    non_salary_limit: Money = Decimal("0.40")
    service_bonus_year_days: int = 360
    service_bonus_top: Money = Decimal("0")
    service_bonus_rate_below_top: Money = Decimal("0.50")
    service_bonus_rate_above_top: Money = Decimal("0.35")
    arl_risk_class: str = "I"
    health_exempt: bool = False
    sena_exempt: bool = False
    icbf_exempt: bool = False

    def validate(self):
        rates = (
            "employee_health_rate", "employee_pension_rate",
            "service_bonus_rate", "severance_rate", "severance_interest_rate",
            "christmas_bonus_rate", "vacation_rate", "vacation_bonus_rate",
            "employer_pension_rate", "employer_health_rate", "employer_health_exempt_rate",
            "compensation_fund_rate", "sena_rate", "sena_exempt_rate", "icbf_rate",
            "icbf_exempt_rate", "non_salary_limit", "service_bonus_rate_below_top",
            "service_bonus_rate_above_top",
        )
        for name in rates:
            value = money(getattr(self, name))
            if value < ZERO:
                raise PayrollDomainError(f"{name} cannot be negative")
        if not ZERO <= money(self.non_salary_limit) <= ONE:
            raise PayrollDomainError("non_salary_limit must be between 0 and 1")
        if self.vacation_divisor <= ZERO:
            raise PayrollDomainError("vacation_divisor must be positive")
        if self.service_bonus_year_days <= 0:
            raise PayrollDomainError("service_bonus_year_days must be positive")
        if self.arl_risk_class not in self.arl_rates:
            raise PayrollDomainError("arl risk class has no configured rate")
        if money(self.arl_rates[self.arl_risk_class]) < ZERO:
            raise PayrollDomainError("ARL rate cannot be negative")


@dataclass(frozen=True)
class PayrollLine:
    code: str
    amount: Money
    is_salary: bool
    affects_ibc: bool


@dataclass(frozen=True)
class PayrollResult:
    employee_id: str
    employee_type: str
    period: str
    days_worked: int
    base_salary: Money
    salary_adjustments: Money
    salary_concepts: tuple[PayrollLine, ...]
    non_salary_concepts: tuple[PayrollLine, ...]
    gross_salary: Money
    non_salary_total: Money
    ibc: Money
    employee_health: Money
    employee_pension: Money
    employee_other_deductions: Money
    total_employee_deductions: Money
    service_bonus_provision: Money
    severance_provision: Money
    severance_interest: Money
    christmas_bonus_provision: Money
    vacation_provision: Money
    vacation_bonus_provision: Money
    service_bonus: Money
    employer_pension: Money
    employer_health: Money
    arl: Money
    compensation_fund: Money
    sena: Money
    icbf: Money
    total_employer_contributions: Money
    net_salary: Money
    total_employer_cost: Money

    def to_dict(self):
        result = asdict(self)
        result["salary_concepts"] = [asdict(line) for line in self.salary_concepts]
        result["non_salary_concepts"] = [asdict(line) for line in self.non_salary_concepts]
        return result


def _salary_base(employee: PayrollEmployee) -> Money:
    if employee.employee_type == "Professor":
        if money(employee.base_monthly_salary) > ZERO:
            return money(employee.base_monthly_salary)
        if employee.employment_type.strip().lower() == "catedratico" and money(employee.hourly_rate) > ZERO:
            return money(employee.hourly_rate) * money(employee.hours_worked)
        return employee.total_points * money(employee.point_value)
    return money(employee.base_monthly_salary)


def calculate_payroll(
    employee: PayrollEmployee,
    period: PayrollPeriod,
    rules: PayrollRules,
    novelties: Iterable[PayrollNovelty] = (),
    other_deductions: Money = ZERO,
) -> PayrollResult:
    """Calculates one employee without I/O, mutation, GUI, or persistence."""
    employee.validate()
    period.validate()
    rules.validate()
    novelty_list = tuple(novelties)
    for novelty in novelty_list:
        novelty.validate()
    other_deductions = _non_negative(other_deductions, "other_deductions")

    base_salary = _salary_base(employee)
    salary_adjustments = sum(
        (money(n.amount) * money(n.quantity) for n in novelty_list if n.is_salary), ZERO
    )
    salary_lines = [PayrollLine("SALARY_BASE", base_salary, True, True)]
    non_salary_lines = []
    for novelty in novelty_list:
        amount = money(novelty.amount) * money(novelty.quantity)
        line = PayrollLine(novelty.code, amount, novelty.is_salary, novelty.affects_ibc)
        (salary_lines if novelty.is_salary else non_salary_lines).append(line)
    constitutive_salary = base_salary + salary_adjustments
    non_salary_total = sum((line.amount for line in non_salary_lines), ZERO)
    non_salary_ibc_base = sum(
        (line.amount for line in non_salary_lines if line.affects_ibc), ZERO
    )
    gross_salary = constitutive_salary + non_salary_total

    limit = money(rules.non_salary_limit)
    allowed_non_salary = round_peso(gross_salary * limit)
    excess_non_salary = max(ZERO, non_salary_total - allowed_non_salary)
    ibc = constitutive_salary + non_salary_ibc_base + excess_non_salary

    employee_health_rate = ZERO if rules.health_exempt else money(rules.employee_health_rate)
    employee_pension_rate = money(rules.employee_pension_rate)
    employee_health = round_peso(ibc * employee_health_rate)
    employee_pension = round_peso(ibc * employee_pension_rate)
    total_employee_deductions = employee_health + employee_pension + round_peso(other_deductions)

    adjusted_base = constitutive_salary * money(period.days_worked) / DAYS_IN_YEAR
    service_bonus_provision = round_peso(adjusted_base * money(rules.service_bonus_rate))
    severance_provision = round_peso(adjusted_base * money(rules.severance_rate))
    severance_interest = round_peso(severance_provision * money(period.days_worked) * money(rules.severance_interest_rate) / DAYS_IN_YEAR)
    christmas_bonus_provision = round_peso(adjusted_base * money(rules.christmas_bonus_rate))
    vacation_provision = round_peso(constitutive_salary * money(period.days_worked) * money(rules.vacation_rate) / money(rules.vacation_divisor))
    vacation_bonus_provision = round_peso(adjusted_base * money(rules.vacation_bonus_rate))

    service_bonus = ZERO
    if employee.continuous_service_days >= rules.service_bonus_year_days:
        bonus_rate = (
            money(rules.service_bonus_rate_below_top)
            if money(employee.base_monthly_salary) <= money(rules.service_bonus_top)
            else money(rules.service_bonus_rate_above_top)
        )
        service_bonus = round_peso(money(employee.base_monthly_salary) * bonus_rate)

    arl = round_peso(ibc * money(rules.arl_rates[rules.arl_risk_class]))
    employer_pension = round_peso(ibc * money(rules.employer_pension_rate))
    employer_health = round_peso(ibc * (money(rules.employer_health_exempt_rate) if rules.health_exempt else money(rules.employer_health_rate)))
    compensation_fund = round_peso(ibc * money(rules.compensation_fund_rate))
    sena = round_peso(ibc * (money(rules.sena_exempt_rate) if rules.sena_exempt else money(rules.sena_rate)))
    icbf = round_peso(ibc * (money(rules.icbf_exempt_rate) if rules.icbf_exempt else money(rules.icbf_rate)))
    total_employer_contributions = sum((employer_pension, employer_health, arl, compensation_fund, sena, icbf), ZERO)

    legal_benefits = service_bonus_provision + severance_provision + severance_interest + christmas_bonus_provision + vacation_provision + vacation_bonus_provision
    net_salary = gross_salary + service_bonus - total_employee_deductions
    total_employer_cost = gross_salary + service_bonus + legal_benefits + total_employer_contributions

    return PayrollResult(
        employee_id=str(employee.employee_id), employee_type=employee.employee_type,
        period=period.period, days_worked=period.days_worked, base_salary=round_peso(base_salary),
        salary_adjustments=round_peso(salary_adjustments), salary_concepts=tuple(salary_lines),
        non_salary_concepts=tuple(non_salary_lines), gross_salary=round_peso(gross_salary),
        non_salary_total=round_peso(non_salary_total), ibc=round_peso(ibc),
        employee_health=employee_health, employee_pension=employee_pension,
        employee_other_deductions=round_peso(other_deductions),
        total_employee_deductions=round_peso(total_employee_deductions),
        service_bonus_provision=service_bonus_provision, severance_provision=severance_provision,
        severance_interest=severance_interest, christmas_bonus_provision=christmas_bonus_provision,
        vacation_provision=vacation_provision, vacation_bonus_provision=vacation_bonus_provision,
        service_bonus=service_bonus, employer_pension=employer_pension,
        employer_health=employer_health, arl=arl, compensation_fund=compensation_fund,
        sena=sena, icbf=icbf, total_employer_contributions=total_employer_contributions,
        net_salary=round_peso(net_salary), total_employer_cost=round_peso(total_employer_cost),
    )


def professor_snapshot(professor) -> PayrollEmployee:
    employment_type = professor.employment_type
    if getattr(professor, "salary_type", "").upper() in {"HOURLY", "HOURLY_RATE"}:
        employment_type = "catedratico"
    return PayrollEmployee(
        employee_id=str(professor.professor_id), employee_type="Professor",
        employment_type=employment_type, base_monthly_salary=money(professor.base_monthly_salary),
        point_value=money(professor.point_value), category_score=money(professor.category_score),
        title_score=money(professor.title_score), experience_score=money(professor.experience_score),
        productivity_score=money(professor.productivity_score), academic_management_score=money(professor.academic_management_score),
        hourly_rate=money(getattr(professor, "hourly_rate", 0)),
        hours_worked=money(getattr(professor, "lecture_hours", 0)),
        continuous_service_days=getattr(professor, "continuous_service_days", 0),
        active=professor.active,
    )


def administrative_snapshot(administrative) -> PayrollEmployee:
    return PayrollEmployee(
        employee_id=str(administrative.administrative_id), employee_type="Administrative",
        employment_type=administrative.employment_type, base_monthly_salary=money(administrative.base_salary),
        continuous_service_days=getattr(administrative, "continuous_service_days", 0),
        active=administrative.active,
    )

File: PITA-Taller1/cpp/payroll_engine.cpp
#include "payroll_engine.hpp"

#include <algorithm>
#include <cmath>
#include <ctime>
#include <stdexcept>

namespace pita::payroll {
namespace {

constexpr double DAYS_IN_YEAR = 360.0;

std::string timestamp() {
    const std::time_t current = std::time(nullptr);
    std::tm local{};
    const std::tm* localTime = std::localtime(&current);
    if (localTime != nullptr) local = *localTime;
    char buffer[32]{};
    std::strftime(buffer, sizeof(buffer), "%Y-%m-%dT%H:%M:%S", &local);
    return buffer;
}

Money roundPeso(double value) {
    return static_cast<Money>(std::llround(value));
}

Money multiply(Money value, Money quantity) {
    return value * quantity;
}

double arlRate(const PayrollRules& rules) {
    for (const auto& configured : rules.arlRates) {
        if (configured.first == rules.arlRiskClass) return configured.second;
    }
    throw std::invalid_argument("ARL risk class has no configured rate");
}

void validateRate(double value, const char* name) {
    if (value < 0.0) throw std::invalid_argument(std::string(name) + " cannot be negative");
}

}  // namespace

void validate(const PayrollPeriod& period) {
    if (period.period.empty()) throw std::invalid_argument("period is required");
    if (period.daysWorked < 0 || period.daysWorked > 360)
        throw std::invalid_argument("daysWorked must be between 0 and 360");
}

void validate(const PayrollEmployee& employee) {
    if (employee.employeeId.empty()) throw std::invalid_argument("employeeId is required");
    if (employee.employeeType != "Professor" && employee.employeeType != "Administrative")
        throw std::invalid_argument("unsupported employeeType");
    for (const auto value : {employee.baseMonthlySalary, employee.pointValue, employee.hourlyRate, employee.hoursWorked}) {
        if (value < 0) throw std::invalid_argument("employee monetary values cannot be negative");
    }
    for (const auto value : {employee.categoryScore, employee.titleScore, employee.experienceScore,
                             employee.productivityScore, employee.academicManagementScore}) {
        validateRate(value, "academic score");
    }
    if (employee.continuousServiceDays < 0)
        throw std::invalid_argument("continuousServiceDays cannot be negative");
}

void validate(const PayrollRules& rules) {
    for (const auto configured : {
             rules.employeeHealthRate, rules.employeePensionRate, rules.serviceBonusRate,
             rules.severanceRate, rules.severanceInterestRate, rules.christmasBonusRate,
             rules.vacationRate, rules.vacationBonusRate, rules.employerPensionRate,
             rules.employerHealthRate, rules.employerHealthExemptRate, rules.compensationFundRate,
             rules.senaRate, rules.senaExemptRate, rules.icbfRate, rules.icbfExemptRate,
             rules.serviceBonusRateBelowTop, rules.serviceBonusRateAboveTop}) {
        if (configured < 0.0) throw std::invalid_argument("payroll rate cannot be negative");
    }
    if (rules.vacationDivisor <= 0.0) throw std::invalid_argument("vacationDivisor must be positive");
    if (rules.serviceBonusYearDays <= 0) throw std::invalid_argument("serviceBonusYearDays must be positive");
    if (rules.nonSalaryLimit < 0.0 || rules.nonSalaryLimit > 1.0)
        throw std::invalid_argument("nonSalaryLimit must be between 0 and 1");
    arlRate(rules);
}

double totalPoints(const PayrollEmployee& employee) {
    return employee.categoryScore + employee.titleScore + employee.experienceScore +
           employee.productivityScore + employee.academicManagementScore;
}

Money salaryBase(const PayrollEmployee& employee) {
    if (employee.employeeType == "Professor") {
        if (employee.baseMonthlySalary > 0) return employee.baseMonthlySalary;
        if (employee.employmentType == "catedratico" && employee.hourlyRate > 0)
            return employee.hourlyRate * employee.hoursWorked;
        return roundPeso(totalPoints(employee) * static_cast<double>(employee.pointValue));
    }
    return employee.baseMonthlySalary;
}

PayrollResult calculatePayroll(
    const PayrollEmployee& employee,
    const PayrollPeriod& period,
    const PayrollRules& rules,
    const std::vector<PayrollNovelty>& novelties,
    Money otherDeductions) {
    validate(employee);
    validate(period);
    validate(rules);
    if (otherDeductions < 0) throw std::invalid_argument("otherDeductions cannot be negative");

    const Money base = salaryBase(employee);
    Money salaryAdjustments = 0;
    Money nonSalaryTotal = 0;
    Money nonSalaryIbcBase = 0;
    std::vector<PayrollLine> salaryConcepts{{"SALARY_BASE", base, true, true, "PayrollEngine", "SALARY_BASE"}};
    std::vector<PayrollLine> nonSalaryConcepts;

    for (const auto& novelty : novelties) {
        if (novelty.code.empty() || novelty.amount < 0 || novelty.quantity < 0)
            throw std::invalid_argument("invalid payroll novelty");
        const Money amount = multiply(novelty.amount, novelty.quantity);
        PayrollLine line{novelty.code, amount, novelty.isSalary, novelty.affectsIbc, "PayrollNovelty", novelty.code};
        if (novelty.isSalary) {
            salaryAdjustments += amount;
            salaryConcepts.push_back(line);
        } else {
            nonSalaryTotal += amount;
            if (novelty.affectsIbc) nonSalaryIbcBase += amount;
            nonSalaryConcepts.push_back(line);
        }
    }

    const Money constitutiveSalary = base + salaryAdjustments;
    const Money grossSalary = constitutiveSalary + nonSalaryTotal;
    const Money allowedNonSalary = roundPeso(static_cast<double>(grossSalary) * rules.nonSalaryLimit);
    const Money excessNonSalary = std::max<Money>(0, nonSalaryTotal - allowedNonSalary);
    const Money ibc = constitutiveSalary + nonSalaryIbcBase + excessNonSalary;

    const Money employeeHealth = roundPeso(static_cast<double>(ibc) * (rules.healthExempt ? 0.0 : rules.employeeHealthRate));
    const Money employeePension = roundPeso(static_cast<double>(ibc) * rules.employeePensionRate);
    const Money totalEmployeeDeductions = employeeHealth + employeePension + otherDeductions;

    const double adjustedBase = static_cast<double>(constitutiveSalary) * period.daysWorked / DAYS_IN_YEAR;
    const Money serviceBonusProvision = roundPeso(adjustedBase * rules.serviceBonusRate);
    const Money severanceProvision = roundPeso(adjustedBase * rules.severanceRate);
    const Money severanceInterest = roundPeso(static_cast<double>(severanceProvision) * period.daysWorked * rules.severanceInterestRate / DAYS_IN_YEAR);
    const Money christmasBonusProvision = roundPeso(adjustedBase * rules.christmasBonusRate);
    const Money vacationProvision = roundPeso(static_cast<double>(constitutiveSalary) * period.daysWorked * rules.vacationRate / rules.vacationDivisor);
    const Money vacationBonusProvision = roundPeso(adjustedBase * rules.vacationBonusRate);

    Money serviceBonus = 0;
    if (employee.continuousServiceDays >= rules.serviceBonusYearDays) {
        const double rate = employee.baseMonthlySalary <= rules.serviceBonusTop
            ? rules.serviceBonusRateBelowTop : rules.serviceBonusRateAboveTop;
        serviceBonus = roundPeso(static_cast<double>(employee.baseMonthlySalary) * rate);
    }

    const Money employerPension = roundPeso(static_cast<double>(ibc) * rules.employerPensionRate);
    const Money employerHealth = roundPeso(static_cast<double>(ibc) * (rules.healthExempt ? rules.employerHealthExemptRate : rules.employerHealthRate));
    const Money arl = roundPeso(static_cast<double>(ibc) * arlRate(rules));
    const Money compensationFund = roundPeso(static_cast<double>(ibc) * rules.compensationFundRate);
    const Money sena = roundPeso(static_cast<double>(ibc) * (rules.senaExempt ? rules.senaExemptRate : rules.senaRate));
    const Money icbf = roundPeso(static_cast<double>(ibc) * (rules.icbfExempt ? rules.icbfExemptRate : rules.icbfRate));
    const Money totalEmployerContributions = employerPension + employerHealth + arl + compensationFund + sena + icbf;
    const Money legalBenefits = serviceBonusProvision + severanceProvision + severanceInterest + christmasBonusProvision + vacationProvision + vacationBonusProvision;

    PayrollResult result;
    result.employeeId = employee.employeeId;
    result.employeeType = employee.employeeType;
    result.period = period.period;
    result.daysWorked = period.daysWorked;
    result.baseSalary = base;
    result.salaryAdjustments = salaryAdjustments;
    result.salaryConcepts = std::move(salaryConcepts);
    result.nonSalaryConcepts = std::move(nonSalaryConcepts);
    result.grossSalary = grossSalary;
    result.nonSalaryTotal = nonSalaryTotal;
    result.ibc = ibc;
    result.employeeHealth = employeeHealth;
    result.employeePension = employeePension;
    result.employeeOtherDeductions = otherDeductions;
    result.totalEmployeeDeductions = totalEmployeeDeductions;
    result.serviceBonusProvision = serviceBonusProvision;
    result.severanceProvision = severanceProvision;
    result.severanceInterest = severanceInterest;
    result.christmasBonusProvision = christmasBonusProvision;
    result.vacationProvision = vacationProvision;
    result.vacationBonusProvision = vacationBonusProvision;
    result.serviceBonus = serviceBonus;
    result.employerPension = employerPension;
    result.employerHealth = employerHealth;
    result.arl = arl;
    result.compensationFund = compensationFund;
    result.sena = sena;
    result.icbf = icbf;
    result.totalEmployerContributions = totalEmployerContributions;
    result.employerContributions = {
        {"EMPLOYER_PENSION", employerPension, rules.employerPensionRate},
        {"EMPLOYER_HEALTH", employerHealth, rules.healthExempt ? rules.employerHealthExemptRate : rules.employerHealthRate},
        {"ARL", arl, arlRate(rules)},
        {"COMPENSATION_FUND", compensationFund, rules.compensationFundRate},
        {"SENA", sena, rules.senaExempt ? rules.senaExemptRate : rules.senaRate},
        {"ICBF", icbf, rules.icbfExempt ? rules.icbfExemptRate : rules.icbfRate},
    };
    result.netSalary = grossSalary + serviceBonus - totalEmployeeDeductions;
    result.totalEmployerCost = grossSalary + serviceBonus + legalBenefits + totalEmployerContributions;
    result.calculationTrace = {
        "netSalary=grossSalary+serviceBonus-totalEmployeeDeductions",
        "ibc=constitutiveSalary+nonSalaryIbcBase+excessNonSalaryOver40Percent",
        "totalEmployerCost=grossSalary+serviceBonus+legalBenefits+totalEmployerContributions",
    };
    return result;
}

Payslip generatePayslip(
    const PayrollEmployee& employee,
    const PayrollPeriod& period,
    const PayrollResult& result,
    const PayrollConfiguration& configuration,
    const std::string& runId) {
    validate(employee);
    validate(period);
    if (result.employeeId != employee.employeeId || result.period != period.period)
        throw std::invalid_argument("PayrollResult does not belong to employee and period");
    Payslip payslip;
    payslip.payslipId = "PS-" + employee.employeeId + "-" + period.period;
    payslip.runId = runId;
    payslip.periodId = period.period;
    payslip.liquidationDate = timestamp();
    payslip.institution = configuration;
    payslip.employee = employee;
    payslip.result = result;
    payslip.arlRiskClass = "I";
    payslip.traceId = runId + ":" + employee.employeeId;
    return payslip;
}

}  // namespace pita::payroll
